compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <random>
+
+#include "iree/compiler/Reducer/Strategies/DeltaStrategies.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler;
+
+void mlir::iree_compiler::reduceLinalgOnTensorsDelta(ChunkManager &chunker,
+                                                     WorkItem &workItem) {
+  ModuleOp module = workItem.getModule();
+
+  SmallVector<linalg::LinalgOp> linalgOps;
+  module.walk([&](linalg::LinalgOp op) {
+    if (!op.hasTensorSemantics())
+      return;
+    // Op should have at least on tensor input, otherwise it is not worth
+    // reducing.
+    bool hasAtleastOneTensorInput = false;
+    for (auto *input : op.getDpsInputOperands()) {
+      if (isa<RankedTensorType>(input->get().getType())) {
+        hasAtleastOneTensorInput = true;
+        break;
+      }
+    }
+
+    if (!hasAtleastOneTensorInput)
+      return;
+
+    // There should be only 1 tensor output.
+    if (op.getNumDpsInits() != 1)
+      return;
+    if (!isa<RankedTensorType>(op.getDpsInitOperand(0)->get().getType()))
+      return;
+
+    if (!chunker.shouldFeatureBeKept()) {
+      linalgOps.push_back(op);
+    }
+  });
+
+  if (linalgOps.empty()) {
+    return;
+  }
+
+  OpBuilder builder = workItem.getBuilder();
+  for (auto linalgOp : linalgOps) {
+    builder.setInsertionPoint(linalgOp);
+    Value out = linalgOp->getResult(0);
+    Value init = linalgOp.getDpsInitOperand(0)->get();
+    RankedTensorType outType = cast<RankedTensorType>(out.getType());
+    Value newOut;
+
+    // TODO: Add support for dynamic shapes using ValueBoundsInterface.
```

Comment:
naked TODOs should be rare and instead TODOs should be targeted at github users (yourself) or issues
https://google.github.io/styleguide/cppguide.html#TODO_Comments
(`// TODO(groverkss): do something.` or `// TODO(#issuenumber): do something.`)